x/slashing/keeper/infractions.go

```diff
@@ -36,30 +36,38 @@ func (k Keeper) HandleValidatorSignatureConcurrent(ctx sdk.Context, addr cryptot
 
 	window := k.SignedBlocksWindow(ctx)
 
+	index := signInfo.IndexOffset
+
 	missedInfo, found = k.GetValidatorMissedBlocks(ctx, consAddr)
+
 	if !found {
+		arrLen := (window + 63) / 64
 		missedInfo = types.ValidatorMissedBlockArray{
-			Address:       consAddr.String(),
-			MissedHeights: make([]int64, 0),
+			Address:      consAddr.String(),
+			WindowSize:   window,
+			MissedBlocks: make([]uint64, arrLen),
 		}
 	}
-	cutoff := height - window
-	// go through and increment number evicted as necessary
-	numberEvicted := 0
-	for _, missedHeight := range missedInfo.MissedHeights {
-		if missedHeight <= cutoff {
-			numberEvicted += 1
-		} else {
-			break
-		}
+	if found && missedInfo.WindowSize != window {
+		missedInfo, signInfo, index = k.ResizeMissedBlockArray(missedInfo, signInfo, window, index)
 	}
-	// update Missed Heights by excluding heights outside of the window
-	missedInfo.MissedHeights = missedInfo.MissedHeights[numberEvicted:]
+	previous := k.GetBooleanFromBitGroups(missedInfo.MissedBlocks, index)
 	missed := !signed
-	if missed {
-		missedInfo.MissedHeights = append(missedInfo.MissedHeights, height-1)
+	switch {
+	case !previous && missed:
+		// Array value has changed from not missed to missed, increment counter
+		signInfo.MissedBlocksCounter++
+		missedInfo.MissedBlocks = k.SetGetBooleanInBitGroups(missedInfo.MissedBlocks, index, true)
```

Comment:
nit - last param could just be `missed`

Reply:
we need the previous value to calculate whether the missed counter needs incrementing or decrementing. However we can move the `setBooleanInBitGroups` out of switch case